GoogleForCreators/web-stories-wp
# Media: Convert HEIF images to WebP/JPEG

## Context



Photos taken on iPhone typically have an `.heic` file extension. The HEIF/HEIC image format is great because it's super small, but it's not supported by any browsers.

## Summary



This change implements HEIC image decoding via ~a simple web worker~ a script that loads libheif-js from an external script that then converts the image to WebP via Canvas.

Safari does not support WebP in Canvas though, so there it falls back to JPEG.

## Relevant Technical Choices



`libheif-js` uses LGPL 3.0, which is why we don't bundle it in the plugin but instead load it dynamically.

We can't use it in a worker because of cross-origin isolation limitations.

So the whole decoding happens in the main thread. In my testing it was usually quite fast, so probably not a big deal.

## To-do



* Load `libheif-js` from CDN

## User-facing changes



https://user-images.githubusercontent.com/841956/218522406-8dc493bd-744f-44a7-8850-d4b7aa8f0792.mov

## Testing Instructions






This PR can be tested by following these steps:

1. Try uploading some HEIC images to the editor via drag & drop  
  I used some from https://github.com/alexcorvi/heic2any/tree/master/demo
2. Verify the images are converted to WebP/JPEG and work fine when previewing the story.


## Reviews

### Does this PR have a security-related impact?



No

### Does this PR change what data or activity we track or use?



No

### Does this PR have a legal-related impact?



`libheif-js` uses LGPL 3.0, which is why we don't bundle it in the plugin but instead load it dynamically.

## Checklist




---



Fixes #8292
